src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+          });
+
+        fill_fetch_responses(
+          octx,
+          std::move(results.read_results),
+          fetch.responses,
+          fetch.start_time);
+
+        // Remove errored fetches from fetch.requests and fetch.responses.
+        if (!results.errored_partitions.empty()) {
+            std::sort(
+              results.errored_partitions.begin(),
+              results.errored_partitions.end());
+            size_t shift = 0;
+            for (auto i : results.errored_partitions) {
+                fetch.requests.erase(fetch.requests.begin() + (i - shift));
+                fetch.responses.erase(fetch.responses.begin() + (i - shift));
+                shift += 1;
+            }
+        }
+
+        _last_result_size[shard] = results.total_size;
+        _completed_shard_fetches.push_back(std::move(fetch));
+        _has_data.signal();
+    }
+
+    std::optional<ss::future<>> register_hwm_wait(
+      shard_local_fetch_context& ctx,
+      const model::ktp& ktp,
+      model::offset offset,
+      ss::abort_source& as) {
+        auto part = ctx.mgr.get(ktp);
+        // If the partition can't be found then it's since been moved
+        if (!part) {
+            return {};
+        }
+
+        auto consensus = part->raft();
+        if (!consensus) {
+            return {};
+        }
+
+        return {ignore_exceptions(consensus->visible_offset_monitor().wait(
+          offset, model::no_timeout, as))};
+    }
+
+    /**
+     * The shard fetch worker will first attempt to query every partition in
+     * `fetch` for data.
+     *
+     * If there is enough data to meet `min_fetch_bytes` or if
+     * the fetch coordinator has sent an abort then the worker returns.
+     *
+     * Otherwise it'll wait for the last_visible_index to change on one of the
+     * partitions in `fetch` then attempt to gather enough data again.
+     *
+     * This will repeat until `min_fetch_bytes` is met or an abort is received
+     * from the coordinator.
+     */
+    ss::future<worker_result>
+    shard_fetch_worker(shard_local_fetch_context ctx) {
+        size_t total_size = 0;
+        std::vector<size_t> errored_partitions;
+
+        // The last visible indexes need to be populated before partitions are
+        // read. If they are populated afterwards then the last_visible_index
+        // could be updated after the partition is read, but before the index is
+        // queried. This leads to a race condition where we're waiting for the
+        // last_visible_index to change even though the partition has data we
+        // haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the `commit_offset`
+        // as the commit offset won't change when partitions are produced to
+        // without acks=all. The `last_visible_index` more closely corresponds
+        // to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{ctx.requests.size()};
```

Comment:
We do get a highwater offset back from `fetch_ntps_in_parallel`. However, it is a Kafka translated offset and I'm unsure if we can translated it back to the raft offset. And if we could I'm unsure if it is analogous to the `last_visible_index` in raft. Will take a look into `fetch_ntps_in_parallel` though to see if it can be modified to return the `last_visible_index`.
 
For a specific offset either enough of the partitions are far enough behind the highwater mark(HWM) that the first `fetch_ntps_in_parallel` can read enough data. In this case the `last_visible_indexes` won't matter as we would have no need to wait on partitions to receive data.

In the other case the fetch will have read up the HWM on all partitions without meeting the min data requirement. In this case we only have to wait for the `last_visible_index` of a partition to change to know there is more data to read.

Reply:
I see thanks. 

Maybe have a look whether we can get some info out of fetch_ntps_in_parallel but if that requires some back and forth then it probably just complicates stuff.

Reply:
are we sure its a kafka translated offset? I think at that layer offsets are still raft offsets, and its the responsibility of the `kafka::replicated_partition` to run the results through the offset translator to return kafka offsets to the client

Reply:
+1 let's make sure we're translating offsets correctly. Can we make sure we're using `kafka::replicated_partition` everywhere here? Maybe we need to add translation layer for the `offset_monitor`, i.e. `kafka::partition_offset_monitor` that translates the offsets there.